### PR TITLE
externalresources(engine): Support s3 as storage (Part I)

### PR DESCRIPTION
**Pull request by liuzix**
### What problem does this PR solve?

Issue Number: ref #7114

### What is changed and how it works?
- Implement `FileManager` for s3.

#### Tests 

 - Unit test
 - Integration test

##### Will it cause performance regression or break compatibility?
No.

##### Do you need to update user documentation, design documentation or monitoring documentation?
No.

### Release note 

```release-note
None.
```

### CODE REVIEW

File: engine/pkg/externalresource/s3/s3_file_manager.go [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+			&internal.ResourceFilesNotFoundError{
+				Ident: ident,
+			})
+	}
+
+	return desc, nil
+}
+
+// RemoveTemporaryFiles removes all temporary resources (those that are not persisted).
+// It can only be used to clean up resources created by the local executor.
```
Comment (CharlesCheung96) [inline engine/pkg/externalresource/s3/s3_file_manager.go:125]: Better to check executorID within this function.

Comment (liuzix) [reply engine/pkg/externalresource/s3/s3_file_manager.go:125]: I think the caller can guarantee it. This requirement is here due to the risk of data race in s3.

File: engine/pkg/externalresource/s3/index.go [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+
+	select {
+	case <-ctx.Done():
+		return errors.Trace(ctx.Err())
+	case err := <-done:
+		if err != nil {
+			return err
+		}
+		return nil
+	}
```
Comment (amyangfei) [inline engine/pkg/externalresource/s3/index.go:259]: Could it block here if `indexManager` is closed and context is still alive.

File: engine/pkg/externalresource/s3/index.go [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+		m.indexFiles[ident.WorkerID] = index
+	}
+
+	ok = index.SetPersisted(ident)
+	m.indexMu.Unlock()
+	if !ok {
+		return false, nil
+	}
+
+	if err := m.triggerFlush(ctx, ident.WorkerID); err != nil {
```
Comment (amyangfei) [inline engine/pkg/externalresource/s3/index.go:176]: Each persisted file will trigger one file flush to s3, do we need to introduce some batch mechanism for index file flush?

File: engine/pkg/externalresource/s3/index.go [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+		return nil
+	}
+
+	bytes, err := index.Serialize()
+	m.indexMu.Unlock()
+	if err != nil {
+		return err
+	}
+
+	storage, err := m.createStorageForIndexFile(ctx, internal.ResourceScope{
```
Comment (amyangfei) [inline engine/pkg/externalresource/s3/index.go:282]: Is it necessary to create a new storage each time, could the index manager reuse one storage instance?

File: engine/pkg/externalresource/internal/file_manager.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+	Name resModel.ResourceName
+}
+
+// Scope returns the Scope of the ResourceIdent.
+func (i ResourceIdent) Scope() ResourceScope {
+	return i.ResourceScope
+}
+
+// FileManager abstracts the operations on the underlying storage.
+type FileManager interface {
```
[the remark this answers is not in the document]
Comment (CharlesCheung96) [reply engine/pkg/externalresource/internal/file_manager.go:54]: local flie manager implment the interface in [this pr](https://github.com/pingcap/tiflow/pull/7127).

File: engine/pkg/externalresource/s3/index.go [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+		return nil, errors.Annotate(err, "serialize index file to s3")
+	}
+	return bytes, nil
+}
+
+func (f *indexFile) GetPersistedFileSet() map[model.ResourceName]struct{} {
+	return f.PersistedFileSet
+}
+
+type indexManagerImpl struct {
```
Comment (amyangfei) [inline engine/pkg/externalresource/s3/index.go:104]: Would it better to implement `indexManager` with metastore, which may be more flexible, and don't need to worry about rate limit of S3.

**Comment by amyangfei** [comment]
Closed by #7127 and #7204.

**Pull request closed without merging**